app/coffee/DocumentUpdaterController.coffee

```diff
@@ -59,30 +59,27 @@ module.exports = DocumentUpdaterController =
 				HealthCheckManager.check channel, message.key
 
 	_applyUpdateFromDocumentUpdater: (io, doc_id, update) ->
-		clientList = io.sockets.clients(doc_id)
-		# avoid unnecessary work if no clients are connected
-		if clientList.length is 0
-			return
-		# send updates to clients
-		logger.log doc_id: doc_id, version: update.v, source: update.meta?.source, socketIoClients: (client.id for client in clientList), "distributing updates to clients"
-		seen = {}
-		# send messages only to unique clients (due to duplicate entries in io.sockets.clients)
-		for client in clientList when not seen[client.id]
-			seen[client.id] = true
-			if client.id == update.meta.source
-				logger.log doc_id: doc_id, version: update.v, source: update.meta?.source, "distributing update to sender"
-				client.emit "otUpdateApplied", v: update.v, doc: update.doc
-			else if !update.dup # Duplicate ops should just be sent back to sending client for acknowledgement
-				logger.log doc_id: doc_id, version: update.v, source: update.meta?.source, client_id: client.id, "distributing update to collaborator"
-				client.emit "otUpdateApplied", update
-		if Object.keys(seen).length < clientList.length
-			metrics.inc "socket-io.duplicate-clients", 0.1
-			logger.log doc_id: doc_id, socketIoClients: (client.id for client in clientList), "discarded duplicate clients"
+		source = update.meta?.source
+		if io.sockets.connected.hasOwnProperty(source)
+			emitter = io.sockets.connected[source]
+			logger.log {doc_id, version: update.v, source}, "distributing update to sender"
+			emitter.emit "otUpdateApplied", v: update.v, doc: update.doc
+		else
+			emitter = io
+		return if update.dup
+		logger.log {doc_id, version: update.v, source}, "distributing updates to clients"
+		emitter.to(doc_id).emit "otUpdateApplied", update
```

Comment:
> This code needs some comments to explain it.

I simplified the logic and added the comment in 818a3e509e5906a0d75b994b820cdef06fce68e2

---

> I think we could drop the hasOwnProperty as socket.io itself treats the id as safe to use directly as a property name? This will simplify the code a bit.

Done in be289b66821f1228a9bcdd6a29837c7cd8e62d64